# Project: Wrangle and Analyse Data about We Rate Dog twitter account#

##### Subject of the document: 

In [ ]:
wrangle_act

## Introduction

The purpose of this activity is to collect and analyze data in different formats related to the We rate dog account. 

We start with 3 data sources, namely:

 - twitter-archive-enhanced.csv
 - image-predictions.tsv
 - tweet_json.txt

From these data, we will carry out our analysis activity in a jupyter notebook by proceeding in 3 stages, namely:

- Gathering data
- Assessing data
- Cleaning data

This document presents the code for collecting, evaluating, cleaning, analyzing and visualizing data.

It is necessary to import the following packages:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 

## Gathering data

goal : Collect data from several sources:

- Download manually twitter_archive_enhanced.csv
- Download data programmatically from an url :tweet image predictions 
- Build a dataframe using API and Json data

### 1 Dowload manually twitter_archive_enhanced.csv

In [2]:
# We are asked to download manually. We have attached it to the notebook.
# We import the pandas library to read the csv file.
import pandas as pd
df = pd.read_csv("twitter-archive-enhanced.csv")

In [3]:
# A preview of the df:
pd.options.display.max_colwidth = 250
df.head(3)


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   
2                  NaN                       NaN                        NaN   

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0                13                  10  Phineas  None    None   None  None  
1                13                  10    Tilly  None    None   None  None  
2                12                  10   Archie  None    None   None  None

### 2 Download data programmatically from an url : tweet image predictions

In [4]:
# We use the Requests library to download the data from the url
import requests
url ='https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

# We use the tsv file contained in the url for loading into dataframe
with open('image-predictions.tsv', mode ='wb') as file:
    file.write(response.content)
    
df2 = pd.read_csv('image-predictions.tsv', '\t')

In [5]:
# A preview of the df2:
df2.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

### 3 Build a dataframe using API and Json data

In [6]:
# Necessary library for using API with twitter
import tweepy
# Json library for using json file
import json
# library for including time
from timeit import default_timer as timer

In [7]:
# We must create a developer's account with twitter to obtain the connection keys
auth = tweepy.OAuthHandler('XXXXX', 'XXXXX')
auth.set_access_token('XXXXX', 'XXXXX')

# create API-connection to twitter:
api = tweepy.API(auth_handler = auth,
                 parser = tweepy.parsers.JSONParser(),
                 wait_on_rate_limit = True,
                 wait_on_rate_limit_notify = True)



In [8]:
tweet_ids = df.tweet_id.values
print('Our query will display this number corresponding to the number of tweet IDs in the twitter_archive_enhanced', len(tweet_ids))

Our query will display this number corresponding to the number of tweet IDs in the twitter_archive_enhanced 2356


In [9]:
# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit    !!!!! 
    
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

1: 892420643555336193
Fail
2: 892177421306343426
Fail
3: 891815181378084864
Fail
4: 891689557279858688
Fail
5: 891327558926688256
Fail
6: 891087950875897856
Fail
7: 890971913173991426
Fail
8: 890729181411237888
Fail
9: 890609185150312448
Fail
10: 890240255349198849
Fail
11: 890006608113172480
Fail
12: 889880896479866881
Fail
13: 889665388333682689
Fail
14: 889638837579907072
Fail
15: 889531135344209921
Fail
16: 889278841981685760
Fail
17: 888917238123831296
Fail
18: 888804989199671297
Fail
19: 888554962724278272
Fail
20: 888202515573088257
Fail
21: 888078434458587136
Fail
22: 887705289381826560
Fail
23: 887517139158093824
Fail
24: 887473957103951883
Fail
25: 887343217045368832
Fail
26: 887101392804085760
Fail
27: 886983233522544640
Fail
28: 886736880519319552
Fail
29: 886680336477933568
Fail
30: 886366144734445568
Fail
31: 886267009285017600
Fail
32: 886258384151887873
Fail
33: 886054160059072513
Fail
34: 885984800019947520
Fail
35: 885528943205470208
Fail
36: 885518971528720385
Fail
3

In [10]:
# We want to integrate the tweet id, number of retweet and number of favorites in a dataframe

requirements_var = []

with open('tweet_json.txt', 'r') as json_file:
    for line in json_file:
        json_data = json.loads(line)
        requirements_var.append({'tweetID' : json_data['id'],
                                'retweets' : json_data['retweet_count'],
                                'favorites' : json_data['favorite_count']})
        
df3 = pd.DataFrame(requirements_var, columns =['tweetID','retweets','favorites'])
df3 = df3.reset_index(drop = True)  


In [11]:
# A preview of the df3
df3.head()

Empty DataFrame
Columns: [tweetID, retweets, favorites]
Index: []

## Assessing data

goal: Work inspecting our data set :

- lack of tidiness data
- data quality issues

We have gathered 3 tables related to the We rate dog account. Let's do a visual assessment. Then, a programmatic assessment.

### 1 Assessing data - DF (twitter_archive_enhanced)


In [12]:
# Display df:
df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [13]:
#Let's analyse columns, values and types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [14]:
# let's analyse statistic data
df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [15]:
# we should analyze the ratings because in max we get a numerator of 1776 and a denominator of 170.
df.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

There is only one numerator at 20 and 461 numerator at 10. Also, we have a lot of outliers (Ex:420..)

In [16]:
df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

There are 2 numerator at 20 and 2333 numerator at 10. Also, we have a lot of outliers (Ex:170..)

According to the elements and the statement, we can consider that the marking is carried out with a maximum mark of 10/10. 
Indeed, we have 461 numerators registered at 10 and 2333 registered at 10.

In [17]:
# We can count the numerators other than 10.
len(df.query("rating_numerator != 10"))

1895

In [18]:
# We can count the denominators other than 10.
len(df.query("rating_denominator !=10"))

23

In [19]:
#Let's take a closer look at these inconsistent values: numerator <10
df[['tweet_id','in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp' ,'text', 'rating_numerator', 'rating_denominator']] [df['rating_numerator'] < 10]


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
45    883482846933004288                    NaN                  NaN   
229   848212111729840128                    NaN                  NaN   
315   835152434251116546                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
462   817502432452313088                    NaN                  NaN   
485   814578408554463233                    NaN                  NaN   
599   798682547630837760                    NaN                  NaN   
605   798576900688019456                    NaN                  NaN   
730   781661882474196992                    NaN                  NaN   
745   780092040432480260                    NaN                  NaN   
764   777953400541634568                    NaN                  NaN   
765   777885040357281792                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
814   771014301343748096                    NaN                  NaN   
835   768193404517830656                    NaN                  NaN   
845   766423258543644672                    NaN                  NaN   
859   763183847194451968                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
883   760252756032651264                    NaN                  NaN   
896   759099523532779520                    NaN                  NaN   
906   758041019896193024                    NaN                  NaN   
912   757596066325864448                    NaN                  NaN   
915   757354760399941633                    NaN                  NaN   
936   753375668877008896                    NaN                  NaN   
946   752568224206688256                    NaN                  NaN   
948   752334515931054080                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
964   750506206503038976                    NaN                  NaN   
966   750383411068534784                    NaN                  NaN   
987   749036806121881602                    NaN                  NaN   
...                  ...                    ...                  ...   
2312  666776908487630848                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2315  666691418707132416                    NaN                  NaN   
2316  666649482315059201                    NaN                  NaN   
2317  666644823164719104                    NaN                  NaN   
2319  666447344410484738                    NaN                  NaN   
2320  666437273139982337                    NaN                  NaN   
2322  666430724426358785                    NaN                  NaN   
2323  666428276349472768                    NaN                  NaN   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN         

- For numerators less than 10, we  notice that the first line index 45, the decimals are not supported.
However, 13.5 / 10 is not consistent. But, the fact that the rating numerators are greater than the denominators does not need to be cleaned
- With index 315, we had 0 since it's a plagiarism.

In [20]:
#Let's take a closer look at these inconsistent values: numerator >10
df[['tweet_id','in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp', 'text', 'rating_numerator', 'rating_denominator']] [df['rating_numerator'] > 10]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2201  668645506898350081                    NaN                  NaN   
2212  668587383441514497                    NaN                  NaN   
2213  668567822092664832                    NaN                  NaN   
2217  668528771708952576                    NaN                  NaN   
2221  668480044826800133                    NaN                  NaN   
2225  668286279830867968                    NaN                  NaN   
2228  668256321989451776                    NaN                  NaN   
2234  668190681446379520                    NaN                  NaN   
2244  667886921285246976                    NaN                  NaN   
2250  667832474953625600                    NaN                  NaN   
2252  667801013445750784                    NaN                  NaN   
2257  667728196545200128                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2267  667524857454854144                    NaN                  NaN   
2269  667509364010450944                    NaN                  NaN   
2270  667502640335572993                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2275  667453023279554560                    NaN                  NaN   
2278  667435689202614272                    NaN                  NaN   
2283  667200525029539841                    NaN                  NaN   
2284  667192066997374976                    NaN                  NaN   
2292  667160273090932737                    NaN                  NaN   
2293  667152164079423490                    NaN                  NaN   
2301  667044094246576128                    NaN         

- These are simply inconsistent values.

In [21]:
#Let's take a closer look at these inconsistent values: denominator < 10
df[['tweet_id','in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp', 'text', 'rating_numerator', 'rating_denominator']] [df['rating_denominator'] < 10]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17           26259576.0   
516   810984652412424192                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

      retweeted_status_id  retweeted_status_user_id  \
313                   NaN                       NaN   
516                   NaN                       NaN   
2335                  NaN                       NaN   

     retweeted_status_timestamp  \
313                         NaN   
516                         NaN   
2335                        NaN   

                                                                                                                                                                           text  \
313                                                                          @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
2335                                   This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

      rating_numerator  rating_denominator  
313                960                   0  
516                 24                   7  
2335                 1                   2

- Index 316 displays a response and the text 960/0 is included in the numerator and denominator columns.
- For the rest, it is an extraction problem.

In [22]:
#Let's take a closer look at these inconsistent values: denominator > 10
df[['tweet_id','in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','text', 'rating_numerator', 'rating_denominator']] [df['rating_denominator'] > 10]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
342   832088576586297345           8.320875e+17         3.058208e+07   
433   820690176645140481                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   

      retweeted_status_id  retweeted_status_user_id  \
342                   NaN                       NaN   
433                   NaN                       NaN   
784          7.403732e+17              4.196984e+09   
902                   NaN                       NaN   
1068                  NaN                       NaN   
1120                  NaN                       NaN   
1165                  NaN                       NaN   
1202                  NaN                       NaN   
1228                  NaN                       NaN   
1254                  NaN                       NaN   
1274                  NaN                       NaN   
1351                  NaN                       NaN   
1433                  NaN                       NaN   
1598                  NaN                       NaN   
1634                  NaN                       NaN   
1635                  NaN                       NaN   
1662                  NaN                       NaN   
1663                  NaN                       NaN   
1779                  NaN                       NaN   
1843                  NaN                       NaN   

     retweeted_status_timestamp  \
342                         NaN   
433                         NaN   
784   2016-06-08 02:41:38 +0000   
902                         NaN   
1068                        NaN   
1120                        NaN   
1165                        NaN   
1202                        NaN   
1228                        NaN   
1254                        NaN   
1274                        NaN   
1351                        NaN   
1433                        NaN   
1598                        NaN   
1634                        NaN   
1635                        NaN   
1662                        NaN   
1663                        NaN   
1779                        NaN   
1843                        NaN   

                                                                                                                                                    text  \
342                                                                                                             @docmisterio account started on 11/15/15   
433                                                  The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd   
784         RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP

The notations displayed in the text are aberrant and are included in the columns.

The column values in the status id and user id response are not needed according to the statement. We can remove them.

In [23]:
# let's see duplicates
df[df.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [24]:
# Check names
df.name.sort_values().unique()

array(['Abby', 'Ace', 'Acro', 'Adele', 'Aiden', 'Aja', 'Akumi', 'Al',
       'Albert', 'Albus', 'Aldrick', 'Alejandro', 'Alexander',
       'Alexanderson', 'Alf', 'Alfie', 'Alfy', 'Alice', 'Amber', 'Ambrose',
       'Amy', 'Amélie', 'Anakin', 'Andru', 'Andy', 'Angel', 'Anna',
       'Anthony', 'Antony', 'Apollo', 'Aqua', 'Archie', 'Arlen', 'Arlo',
       'Arnie', 'Arnold', 'Arya', 'Ash', 'Asher', 'Ashleigh', 'Aspen',
       'Astrid', 'Atlas', 'Atticus', 'Aubie', 'Augie', 'Autumn', 'Ava',
       'Axel', 'Bailey', 'Baloo', 'Balto', 'Banditt', 'Banjo', 'Barclay',
       'Barney', 'Baron', 'Barry', 'Batdog', 'Bauer', 'Baxter', 'Bayley',
       'BeBe', 'Bear', 'Beau', 'Beckham', 'Beebop', 'Beemo', 'Bell',
       'Bella', 'Belle', 'Ben', 'Benedict', 'Benji', 'Benny', 'Bentley',
       'Berb', 'Berkeley', 'Bernie', 'Bert', 'Bertson', 'Betty', 'Beya',
       'Biden', 'Bilbo', 'Billl', 'Billy', 'Binky', 'Birf', 'Bisquick',
       'Blakely', 'Blanket', 'Blipson', 'Blitz', 'Bloo', 'Bloop', 'Blu',

A name starts with a capital letter, let's look for names that start with a lowercase

In [25]:
#Name with a lowercase
df[df.name.str.islower()].name.unique()

array(['such', 'a', 'quite', 'not', 'one', 'incredibly', 'mad', 'an',
       'very', 'just', 'my', 'his', 'actually', 'getting', 'this',
       'unacceptable', 'all', 'old', 'infuriating', 'the', 'by',
       'officially', 'life', 'light', 'space'], dtype=object)

It's weird that words are inserted instead of the name. Let's take a look at the text column.

In [26]:
# let's display text column in large with names in lowercase
pd.options.display.max_colwidth = 250
df[['text','name']] [df.name.str.islower()]

text  \
22                                  I've yet to rate a Venezuelan Hover Wiener. This is such an honor. 14/10 paw-inspiring af (IG: roxy.thedoxy) https://t.co/20VrLAA8ba   
56           Here is a pupper approaching maximum borkdrive. Zooming at never before seen speeds. 14/10 paw-inspiring af \n(IG: puffie_the_chow) https://t.co/ghXBIIeQZF   
118                         RT @dog_rates: We only rate dogs. This is quite clearly a smol broken polar bear. We'd appreciate if you only send dogs. Thank you... 12/10…   
169                 We only rate dogs. This is quite clearly a smol broken polar bear. We'd appreciate if you only send dogs. Thank you... 12/10 https://t.co/g2nSyGenG9   
193   Guys, we only rate dogs. This is quite clearly a bulbasaur. Please only send dogs. Thank you... 12/10 human used pet, it's super effective https://t.co/Xc7uj1C64x   
335                                                             There's going to be a dog terminal at JFK Airport. This is not a drill. 10/10  \nhttps://t.co/dp5h9bCwU7   
369                                                             Occasionally, we're sent fantastic stories. This is one of them. 14/10 for Grace https://t.co/bZ4axuH6OK   
542                  We only rate dogs. Please stop sending in non-canines like this Freudian Poof Lion. This is incredibly frustrating... 11/10 https://t.co/IZidSrBvhi   
649                                     Here is a perfect example of someone who has their priorities in order. 13/10 for both owner and Forrest https://t.co/LRyMrU7Wfq   
682                                   RT @dog_rates: Say hello to mad pupper. You know what you did. 13/10 would pet until no longer furustrated https://t.co/u1ulQ5heLX   
759                         RT @dog_rates: This is an East African Chalupa Seal. We only rate dogs. Please only send in dogs. Thank you... 10/10 https://t.co/iHe6liLwWR   
773                         RT @dog_rates: We only rate dogs. Pls stop sending in non-canines like this Mongolian grass snake. This is very frustrating. 11/10 https://…   
801                         Guys this is getting so out of hand. We only rate dogs. This is a Galapagos Speed Panda. Pls only send dogs... 10/10 https://t.co/8lpAGaZRFn   
819                          We only rate dogs. Pls stop sending in non-canines like this Arctic Floof Kangaroo. This is very frustrating. 11/10 https://t.co/qlUDuPoE3d   
822                                                           RT @dog_rates: This is just downright precious af. 12/10 for both pupper and doggo https://t.co/o5J479bZUC   
852                               This is my dog. Her name is Zoey. She knows I've been rating other dogs. She's not happy. 13/10 no bias at all https://t.co/ep1NkYoiwB   
924                         This is one of the most inspirational stories I've ever come across. I have no words. 14/10 for both doggo and owner https://t.co/I5ld3eKD5k   
988                             What jokester sent in a pic without a dog in it? This is not @rock_rates. This is @dog_rates. Thank you ...10/10 https://t.co/nDPaYHrtNX   
992                          That is Quizno. This is his beach. He does not tolerate human shenanigans on his beach. 10/10 reclaim ur land doggo https://t.co/vdr7DaRSa7   
993                         This is one of the most reckless puppers I've ever seen. How she got a license in the first place is beyond me. 6/10 https://t.co/z5bAdtn9kd   
1002                        This is a mighty rare blue-tailed hammer sherk. Human almost lost a limb trying to take these. Be careful guys. 8/10 https://t.co/TGenMeXreW   
1004                          Viewer discretion is advised. This is a terrible attack in progress. Not even in water (tragic af). 4/10 bad sherk https://t.co/L3U0j14N5R   
1017                              This is a carrot. We only rate dogs. Please only send in dogs. You all really should know this by now ...11/10 https://t.co/9e48aPrBm2   
1025                          This i

In [27]:
# Let's count empty names
df.name.value_counts()[0:1]

None    745
Name: name, dtype: int64

It seems that words from the text column are used as a name.
In the text column, we cannot find the names of the dogs. We have to replace with Nan and the same thing for the "None".

In [28]:
# how many dogs like doggo?
doggo = df.doggo.value_counts()[1]

In [29]:
print("We have {} dog stage 'doggo'.".format(doggo))

We have 97 dog stage 'doggo'.


In [30]:
# how many dogs like floofer?
floofer = df.floofer.value_counts()[1]

In [31]:
print("We have {} dog stage 'floofer'.".format(floofer))

We have 10 dog stage 'floofer'.


In [32]:
# how many dogs like pupper?
pupper = df.pupper.value_counts()[1]

In [33]:
print("We have {} dog stage 'pupper'.".format(pupper))

We have 257 dog stage 'pupper'.


In [34]:
# how many dogs like puppo?
puppo = df.puppo.value_counts()[1] 

In [35]:
print("We have {} dog stage 'puppo'.".format(puppo))

We have 30 dog stage 'puppo'.


In [36]:
# how many dogs like doggo and pupper?
doggopupper = len(df.query('doggo == "doggo" and pupper == "pupper"'))

In [37]:
print("we have {} dog stage doggo/pupper.".format(doggopupper))

we have 12 dog stage doggo/pupper.


In [38]:
# how many dogs like doggo and floofer?
doggofloofer = len(df.query('doggo == "doggo" and floofer == "floofer"'))

In [39]:
print("we have {} dog stage doggo/floofer.".format(doggofloofer))

we have 1 dog stage doggo/floofer.


In [40]:
# how many dogs like doggo and puppo?
doggopuppo = len(df.query('doggo == "doggo" and puppo == "puppo"'))

In [41]:
print("we have {} dog stage doggo/puppo.".format(doggopuppo))

we have 1 dog stage doggo/puppo.


In [42]:
# how many dogs like floofer and pupper?
len(df.query('floofer == "floofer" and pupper == "pupper"'))

0

In [43]:
# how many dogs like floofer and puppo?
len(df.query('floofer == "floofer" and puppo == "puppo"'))

0

In [44]:
# how many dogs like pupper and puppo?
len(df.query('pupper == "pupper" and puppo == "puppo"'))

0

We need to create a 'dogstage' column and add the values:
"doggo, floofer, pupper, puppo, 'doggo, pupper', 'doggo, floofer' and 'doggo, puppo'".
Also, we need to convert the "none" to NaN.

##### Assessment Results:

##### Quality

###### Completeness:
- Replace wrong names (lowercase names) by NaN
- Convert dog stage column ('doggo','floofer','pupper','puppo')  empty to Nan.

###### Validity:
- get the http link only in the 'source' column.

###### Accuracy:
Nothing

###### Consistency:
Nothing

##### Tidiness

###### Dirty data:
- Delete duplicate expanded url.

###### Messy data:
- Modify type of columns ('tweet id', 'in_reply_to_status_id' and 'in_reply_to_user_id') and convert nulls value to Nan.
- Modify type of columns ('timestam'p and 'retweeted_status_timestamp') and convert nulls value to Nan.
- Drop rows where columns ('in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id' and 'retweeted_status_user_id') are not null
- Create column for 'Dog stage' and add it values and drop useless columns ('doggo'...)

### 2 Assessing data - df2 (image-predictions)

In [45]:
# Display df2:
df2

tweet_id  \
0     666020888022790149   
1     666029285002620928   
2     666033412701032449   
3     666044226329800704   
4     666049248165822465   
5     666050758794694657   
6     666051853826850816   
7     666055525042405380   
8     666057090499244032   
9     666058600524156928   
10    666063827256086533   
11    666071193221509120   
12    666073100786774016   
13    666082916733198337   
14    666094000022159362   
15    666099513787052032   
16    666102155909144576   
17    666104133288665088   
18    666268910803644416   
19    666273097616637952   
20    666287406224695296   
21    666293911632134144   
22    666337882303524864   
23    666345417576210432   
24    666353288456101888   
25    666362758909284353   
26    666373753744588802   
27    666396247373291520   
28    666407126856765440   
29    666411507551481857   
...                  ...   
2045  886366144734445568   
2046  886680336477933568   
2047  886736880519319552   
2048  886983233522544640   
2049  887101392804085760   
2050  887343217045368832   
2051  887473957103951883   
2052  887517139158093824   
2053  887705289381826560   
2054  888078434458587136   
2055  888202515573088257   
2056  888554962724278272   
2057  888804989199671297   
2058  888917238123831296   
2059  889278841981685760   
2060  889531135344209921   
2061  889638837579907072   
2062  889665388333682689   
2063  889880896479866881   
2064  890006608113172480   
2065  890240255349198849   
2066  890609185150312448   
2067  890729181411237888   
2068  890971913173991426   
2069  891087950875897856   
2070  891327558926688256   
2071  891689557279858688   
2072  891815181378084864   
2073  892177421306343426   
2074  892420643555336193   

                                                                                      jpg_url  \
0                                             https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1                                             https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2                                             https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3                                             https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4                                             https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5                                             https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6                                             https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7                                             https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8                                             https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9                                             https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10                                            https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11                                            https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12                                            https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13                                            https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14                                            https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15                                            https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16                                            https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17                                            https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18                                            https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19                                            https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20                                            https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21                                            https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22                                            ht

##### Programmatic assessment:

In [46]:
#Check columns type and content
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [47]:
#Check summary data
df2.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [48]:
#Check duplicates
df2[df2.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [49]:
#Check content
df2.jpg_url.sort_values()

462     https://pbs.twimg.com/ext_tw_video_thumb/674805331965399040/pu/img/-7bw8niVrgIkLKOW.jpg
1297    https://pbs.twimg.com/ext_tw_video_thumb/675354114423808004/pu/img/qL1R_nGLqa6lmkOx.jpg
480     https://pbs.twimg.com/ext_tw_video_thumb/675354114423808004/pu/img/qL1R_nGLqa6lmkOx.jpg
495     https://pbs.twimg.com/ext_tw_video_thumb/675740268751138818/pu/img/dVaVeFAVT-lk_1ZV.jpg
528     https://pbs.twimg.com/ext_tw_video_thumb/676776408941662209/pu/img/k-6I3YEZAQtYPBXR.jpg
539     https://pbs.twimg.com/ext_tw_video_thumb/676957802976419840/pu/img/dCj-qlXo73A5hf6Q.jpg
554     https://pbs.twimg.com/ext_tw_video_thumb/677644010865999872/pu/img/zVHEMYnJKzq1SauT.jpg
571     https://pbs.twimg.com/ext_tw_video_thumb/678399528077250560/pu/img/BOjUNHRsYLeSo0hl.jpg
588     https://pbs.twimg.com/ext_tw_video_thumb/679111114081370114/pu/img/hFca8BHjRopgD0cM.jpg
619     https://pbs.twimg.com/ext_tw_video_thumb/680440290977693696/pu/img/B900g7b-n-zhnwi5.jpg
682     https://pbs.twimg.com/ext_tw_vid

In [50]:
#Check content P1,P2 and P3
df2.p1.sort_values()

1561                      Afghan_hound
1855                      Afghan_hound
1458                      Afghan_hound
1804                      Afghan_hound
446                  African_crocodile
1371                      African_grey
289                African_hunting_dog
1883                          Airedale
678                           Airedale
425                           Airedale
1319                          Airedale
1646                          Airedale
378                           Airedale
49                            Airedale
603                           Airedale
952                           Airedale
189                           Airedale
582                           Airedale
1465                          Airedale
139     American_Staffordshire_terrier
269     American_Staffordshire_terrier
1837    American_Staffordshire_terrier
550     American_Staffordshire_terrier
909     American_Staffordshire_terrier
1582    American_Staffordshire_terrier
758     American_Stafford

In [51]:
#Check missing values
df2[df2.jpg_url.isnull()]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

##### Assessment Results:

##### Quality

###### Completeness:
df (twitter-archive-enhanced)
- Replace wrong names (lowercase names) by NaN
- Convert dog stage column ('doggo','floofer','pupper','puppo')  empty to Nan.

###### Validity:
df (twitter-archive-enhanced)
- get the http link only in the 'source' column.

###### Accuracy:
Nothing

###### Consistency:
df2 (image-predictions)
- Capitalize names for each prediction dog breed

##### Tidiness

###### Dirty data:
df (twitter-archive-enhanced)
- Delete duplicate expanded url.

###### Messy data:
df (twitter-archive-enhanced)
- Modify type of columns ('tweet id', 'in_reply_to_status_id' and 'in_reply_to_user_id') and convert nulls value to Nan.
- Modify type of columns ('timestam'p and 'retweeted_status_timestamp') and convert nulls value to Nan.
- Drop rows where columns ('in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id' and 'retweeted_status_user_id') are not null
- Create column for 'Dog stage' and add it values and drop useless columns ('doggo'...)

df2 (image-predictions)
- Modify type of columns ('tweet id', 'jpg_url' and	'img_num') and convert nulls value to Nan.
- Rename columns for better understanding
- Merge df2 in df


###  3 Assessing data - df3 (tweet_json)

In [52]:
# Display df3:
df3

Empty DataFrame
Columns: [tweetID, retweets, favorites]
Index: []

##### Programmatic assessment:

In [53]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Data columns (total 3 columns):
tweetID      0 non-null object
retweets     0 non-null object
favorites    0 non-null object
dtypes: object(3)
memory usage: 76.0+ bytes


In [54]:
df3.describe()

tweetID  retweets  favorites
count         0         0          0
unique        0         0          0

In [55]:
# check duplicates
df3[df3.duplicated()]

Empty DataFrame
Columns: [tweetID, retweets, favorites]
Index: []

##### Assessment Results:

##### Quality

###### Completeness:
df (twitter-archive-enhanced)
- Replace wrong names (lowercase names) by NaN
- Convert dog stage column ('doggo','floofer','pupper','puppo')  empty to Nan.

###### Validity:
df (twitter-archive-enhanced)
- get the http link only in the 'source' column.

###### Accuracy:
Nothing

###### Consistency:
df2 (image-predictions)
- Capitalize names for each prediction dog breed

##### Tidiness

###### Dirty data:
df (twitter-archive-enhanced)
- Delete duplicate expanded url.

###### Messy data:
df (twitter-archive-enhanced)
- Modify type of columns ('tweet id', 'in_reply_to_status_id' and 'in_reply_to_user_id') and convert nulls value to Nan.
- Modify type of columns ('timestamp' and 'retweeted_status_timestamp') and convert nulls value to Nan.
- Drop rows where columns ('in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id' and 'retweeted_status_user_id') are not null
- Create column for 'Dog stage' and add it values and drop useless columns ('doggo'...)

df2 (image-predictions)
- Modify type of columns ('tweet id', 'jpg_url' and	'img_num') and convert nulls value to Nan.
- Rename columns for better understanding
- Merge df2 in df

df3 (tweet_json)
- type of column ('tweet id')
- Rename columns('retweets' and 'favorites' for better understanding)
- Merge df3 in df

## Cleaning data

goal: It is where we fix the quality and tidiness issues that we identified in the assess step.

we will follow the next step:

- Define
- Code
- Test

In [56]:
# Copy of each piece of data
df_clean = df.copy()
df2_clean = df2.copy()
df3_clean = df3.copy()

#### Define
df (twitter-archive-enhanced)
- Replace wrong names (lowercase names) by NaN
- Replace None by NaN

#### Code

In [57]:
error_names = df_clean[df_clean.name.str.islower()].name.unique()
error_names

array(['such', 'a', 'quite', 'not', 'one', 'incredibly', 'mad', 'an',
       'very', 'just', 'my', 'his', 'actually', 'getting', 'this',
       'unacceptable', 'all', 'old', 'infuriating', 'the', 'by',
       'officially', 'life', 'light', 'space'], dtype=object)

In [58]:
df_clean.name = df_clean.name.replace(error_names, np.nan)

In [59]:
df_clean.name = df_clean.name.replace('None', np.nan)

In [60]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          1502 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

#### Test

In [61]:
df_clean.name.unique()

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', nan, 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver', 'Jim',
       'Zeke', 'Ralphus', 'Canela', 'Gerald', 'Jeffrey', 'Maya', 'Mingus',
       'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey', 'Lilly', 'Earl',
       'Lola', 'Kevin', 'Yogi', 'Noah', 'Bella', 'Grizzwald', 'Rusty',
       'Gus', 'Stanley', 'Alfy', 'Koko', 'Rey', 'Gary', 'Elliot', 'Louis',
       'Jesse', 'Romeo', 'Bailey', 'Duddles', 'Jack', 'Emmy', 'Steven',
       'Beau', 'Snoopy', 'Shadow', 'Terrance', 'Aja', 'Penny', 'Dante',
       'Nelly', 'Ginger', 'Benedict', 'Venti', 'Goose', 'Nugget', 'Cash',
       'Coco', 'Jed', 'Sebastian', 'Walter', 'Sierra', 'Monkey', 'Harry',
       'Kody', 'Lassie', 'Rover', 'Napolean', 'Dawn', 'Boomer', 'Cody',
       'Rumble', 'Clifford', 'Dewey', 'Scout', 'Gizmo', 'Cooper', 'Harold',
       'Shikha', 'Jamesy', 'Lili', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Neptune', 'Quinn', 'Belle', 'Zooey',

#### Define
df (twitter-archive-enhanced)
- Convert dog stage column ('doggo','floofer','pupper','puppo')  'None' to ' '.

#### Code

In [62]:
df_clean.doggo.replace('None', '', inplace = True)
df_clean.floofer.replace('None', '', inplace=True)
df_clean.pupper.replace('None', '',inplace = True)
df_clean.puppo.replace('None', '',inplace = True)

#### Test

In [63]:
df_clean.doggo.value_counts()

         2259
doggo      97
Name: doggo, dtype: int64

In [64]:
df_clean.floofer.value_counts()

           2346
floofer      10
Name: floofer, dtype: int64

In [65]:
df_clean.puppo.value_counts()

         2326
puppo      30
Name: puppo, dtype: int64

#### Define
df (twitter-archive-enhanced)
- get the http link only in the 'source' column.

#### Code

In [66]:
df_clean.source.head()

0    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
1    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
2    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
3    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
4    <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
Name: source, dtype: object

In [67]:
part = df_clean.source.str.split('"', expand= True)
part.head()

0                                   1      2         3  \
0  <a href=  http://twitter.com/download/iphone   rel=  nofollow   
1  <a href=  http://twitter.com/download/iphone   rel=  nofollow   
2  <a href=  http://twitter.com/download/iphone   rel=  nofollow   
3  <a href=  http://twitter.com/download/iphone   rel=  nofollow   
4  <a href=  http://twitter.com/download/iphone   rel=  nofollow   

                         4  
0  >Twitter for iPhone</a>  
1  >Twitter for iPhone</a>  
2  >Twitter for iPhone</a>  
3  >Twitter for iPhone</a>  
4  >Twitter for iPhone</a>

In [68]:
df_clean.source = part[1]

#### Test

In [69]:
df_clean.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp                              source  \
0  2017-08-01 16:23:56 +0000  http://twitter.com/download/iphone   
1  2017-08-01 00:17:27 +0000  http://twitter.com/download/iphone   
2  2017-07-31 00:18:03 +0000  http://twitter.com/download/iphone   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   
2                  NaN                       NaN                        NaN   

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0                13                  10  Phineas                             
1                13                  10    Tilly                             
2                12                  10   Archie

#### Define
df2 (image-predictions)
- Capitalize names for each prediction dog breed

#### Code

In [70]:
prediction_breed1 = df2_clean['p1'].str.capitalize()

In [71]:
df2_clean['p1'] = prediction_breed1

In [72]:
prediction_breed2 = df2_clean['p2'].str.capitalize()

In [73]:
df2_clean['p2'] = prediction_breed2

In [74]:
prediction_breed3 = df2_clean['p3'].str.capitalize()

In [75]:
df2_clean['p3'] = prediction_breed3

#### Test

In [76]:
df2_clean.sample(3)

tweet_id                                          jpg_url  \
1948  862831371563274240  https://pbs.twimg.com/media/C_lkieeVwAAm0L4.jpg   
1461  778286810187399168  https://pbs.twimg.com/media/Cs0HuUTWcAUpSE8.jpg   
444   674644256330530816  https://pbs.twimg.com/media/CVzRXmXWIAA0Fkr.jpg   

      img_num                  p1   p1_conf  p1_dog             p2   p2_conf  \
1948        2  Australian_terrier  0.207281    True  Irish_terrier  0.156296   
1461        1         Boston_bull  0.322070    True            Pug  0.229903   
444         1         Soccer_ball  0.398102   False         Basset  0.335692   

      p2_dog               p3   p3_conf  p3_dog  
1948    True  German_shepherd  0.123536    True  
1461    True           Muzzle  0.101420   False  
444     True   Cocker_spaniel  0.072941    True

#### Define
df (twitter-archive-enhanced)
- Delete duplicate expanded urls.

#### Code

In [77]:
df_clean.expanded_urls.value_counts()

https://www.gofundme.com/3yd6y1c,https://twitter.com/dog_rates/status/878281511006478336/photo/1                                                                                                                                                                                                             2
https://twitter.com/dog_rates/status/750719632563142656/photo/1                                                                                                                                                                                                                                              2
https://twitter.com/dog_rates/status/680055455951884288/photo/1                                                                                                                                                                                                                                              2
https://twitter.com/dog_rates/status/809220051211603969/photo/1,https://twitter.com/dog_rat

In [78]:
urls_clean = df_clean.expanded_urls.str.split(',', expand = True)
urls_clean[0]
df_clean.expanded_urls = urls_clean[0]

#### Test

In [79]:
df_clean.expanded_urls.value_counts()

https://twitter.com/dog_rates/status/762699858130116608/photo/1    2
https://vine.co/v/5QWd3LZqXxd                                      2
https://twitter.com/dog_rates/status/669000397445533696/photo/1    2
https://vine.co/v/ea0OwvPTx9l                                      2
https://twitter.com/dog_rates/status/773308824254029826/photo/1    2
https://twitter.com/dog_rates/status/830583320585068544/photo/1    2
https://twitter.com/dog_rates/status/820749716845686786/photo/1    2
https://twitter.com/dog_rates/status/816450570814898180/photo/1    2
https://twitter.com/dog_rates/status/868880397819494401/photo/1    2
https://twitter.com/dog_rates/status/742423170473463808/photo/1    2
https://twitter.com/dog_rates/status/789986466051088384/photo/1    2
https://twitter.com/dog_rates/status/667152164079423490/photo/1    2
https://twitter.com/dog_rates/status/844704788403113984/photo/1    2
https://www.gofundme.com/my-puppys-double-cataract-surgery         2
https://twitter.com/dog_rates/stat

#### Define
df (twitter-archive-enhanced)
- Modify type of columns ('tweet id', 'in_reply_to_status_id' and 'in_reply_to_user_id') and convert nulls value to Nan.

#### Code

In [80]:
df_clean.tweet_id = df_clean.tweet_id.astype(str)

#### Test

In [81]:
type(df_clean.tweet_id[0])

str

#### Define
df (twitter-archive-enhanced)
- Modify type of columns ('timestamp' and 'retweeted_status_timestamp') and convert nulls value to Nan.

#### Code

In [82]:
df_clean.timestamp = pd.to_datetime(df_clean.timestamp)

In [83]:
df_clean.retweeted_status_timestamp = pd.to_datetime(df_clean.retweeted_status_timestamp)

#### Test

In [84]:
type(df_clean.timestamp[0])

pandas._libs.tslibs.timestamps.Timestamp

In [85]:
type(df_clean.retweeted_status_timestamp[0])

pandas._libs.tslibs.nattype.NaTType

#### Define
df (twitter-archive-enhanced)
- Drop rows where columns ('in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id' and 'retweeted_status_user_id') are not null

#### Code 

In [86]:
df_clean = df_clean[df_clean.in_reply_to_status_id.isnull()]

In [87]:
df_clean = df_clean[df_clean.in_reply_to_user_id.isnull()]

In [88]:
df_clean = df_clean[df_clean.retweeted_status_id.isnull()]

In [89]:
df_clean = df_clean[df_clean.retweeted_status_user_id.isnull()]

In [90]:
df_clean.count()

tweet_id                      2097
in_reply_to_status_id            0
in_reply_to_user_id              0
timestamp                     2097
source                        2097
text                          2097
retweeted_status_id              0
retweeted_status_user_id         0
retweeted_status_timestamp       0
expanded_urls                 2094
rating_numerator              2097
rating_denominator            2097
name                          1390
doggo                         2097
floofer                       2097
pupper                        2097
puppo                         2097
dtype: int64

In [91]:
df_clean.drop(['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'], axis=1, inplace=True)

#### Test

In [92]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null object
timestamp             2097 non-null datetime64[ns]
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  1390 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: datetime64[ns](1), int64(2), object(9)
memory usage: 213.0+ KB


#### Define
df (twitter-archive-enhanced)
- Create column for 'Dog stage' and add it values and drop useless columns ('doggo'...)

#### Code

In [93]:
df_clean['dog_stage'] = df_clean.doggo + df_clean.floofer + df_clean.pupper + df_clean.puppo


df_clean.loc[df_clean.dog_stage == 'z', 'dog_stage'] = 'doggo, pupper'
df_clean.loc[df_clean.dog_stage == 'b', 'dog_stage'] = 'doggo, puppo'
df_clean.loc[df_clean.dog_stage == 'j', 'dog_stage'] = 'doggo, floofer'

In [94]:
# Convert the dog_stage into categorical dtype
df_clean.dog_stage = df_clean.dog_stage.astype('category')

In [95]:
# drop 'doggo', 'floofer', 'pupper', 'puppo' columns
df_clean.drop(['doggo','pupper','puppo','floofer'], axis=1, inplace=True)
df_clean.dog_stage.replace('', np.NaN, inplace=True)

#### Test

In [96]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id              2097 non-null object
timestamp             2097 non-null datetime64[ns]
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  1390 non-null object
dog_stage             336 non-null category
dtypes: category(1), datetime64[ns](1), int64(2), object(5)
memory usage: 149.9+ KB


In [97]:
df_clean.dog_stage.value_counts()

pupper          221
doggo            72
puppo            23
floofer           9
doggopupper       9
doggopuppo        1
doggofloofer      1
                  0
Name: dog_stage, dtype: int64

#### Define
df2 (image-predictions)
- Modify type of columns ('tweet id', 'jpg_url' and	'img_num').

#### Code

In [98]:
df2_clean.tweet_id = df2_clean.tweet_id.astype(str)

In [99]:
df2_clean.jpg_url = df2_clean.jpg_url.astype(str)

In [100]:
df2_clean.img_num = df2_clean.img_num.astype(str)

#### Test

In [101]:
df2_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null object
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), object(6)
memory usage: 152.1+ KB


#### Define
df2 (image-predictions)
- Rename columns for better understanding

#### Code

In [102]:
df2_clean = df2_clean.rename(columns={'p1':'best_breed_prediction', 'p2':'optimal_breed_prediction','p3':'low_breed_prediction', \
                         'p1_conf':'high_rate_predict','p2_conf':'optimal_rate_predict', 'p3_conf':'low_rate_predict'})

#### Test

In [103]:
df2_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id                    2075 non-null object
jpg_url                     2075 non-null object
img_num                     2075 non-null object
best_breed_prediction       2075 non-null object
high_rate_predict           2075 non-null float64
p1_dog                      2075 non-null bool
optimal_breed_prediction    2075 non-null object
optimal_rate_predict        2075 non-null float64
p2_dog                      2075 non-null bool
low_breed_prediction        2075 non-null object
low_rate_predict            2075 non-null float64
p3_dog                      2075 non-null bool
dtypes: bool(3), float64(3), object(6)
memory usage: 152.1+ KB


#### Define
df3 (tweet_json)
- type of column ('tweet id')

#### Code 

In [104]:
df3_clean.tweetID = df3_clean.tweetID.astype(str)

#### Test

In [105]:
type(df3_clean.tweetID[0])

IndexError: index out of bounds

#### Define
df3 (tweet_json)
- Rename columns('retweets' and 'favorites' for better understanding)

#### Code 

In [ ]:
df3_clean = df3_clean.rename(columns={'tweetID':'tweet_id','retweets' : 'retweets_nb', 'favorites':'favorites_nb'})

#### Test 

In [ ]:
df3_clean.info()

#### Define
df3 (tweet_json)
- Change retweets_nb and favorites_nb data types

#### Code

In [ ]:
df3_clean.retweets_nb = df3_clean.retweets_nb.astype(int)
df3_clean.favorites_nb = df3_clean.favorites_nb.astype(int)

#### Test

In [ ]:
type(df3_clean.retweets_nb[0]), type(df3_clean.favorites_nb[0])

In [ ]:
df3_clean.info()

#### Define
df3 (tweet_json)
- Merge df3 in df

#### Code

In [ ]:
df_new = pd.merge(df_clean, df3_clean, on='tweet_id', how='left')

#### Test

In [ ]:
df_new.info()

#### Define
df2 (image-predictions)
- Merge df2 in df_new

#### Code

In [ ]:
df_final = pd.merge(df_new, df2_clean, on='tweet_id', how='left')

#### Test

In [ ]:
df_final.info()

In [ ]:
df_final.describe()

## Save the data in .csv

In [ ]:
df_final.to_csv('twitter_archive_master.csv', encoding='utf-8', index=False )

## Analysis

From the cleaned data, we can determine the most popular breed and the most used names.

In [ ]:
# What are the most used names?
df_final.name.value_counts()[0:10]

In [ ]:
# Which dog breeds define with the best confidence rate (>0.9) ?
df_final.query('high_rate_predict >= 0.9')  ['best_breed_prediction'].value_counts()[0:10]

In [ ]:
# Graphic
import matplotlib.pyplot as plt
df_final.query('high_rate_predict >= 0.9')  ['best_breed_prediction'].value_counts()[0:10].plot(kind='bar');
plt.title('Top 10 best breed', size=22)
plt.xlabel('Dog breed')
plt.ylabel('confidence rate')

plt.plot();

In [ ]:
#What are the 10 most popular dog breeds with ratings greater than or equal to 10?
df_final.query('rating_numerator >= 10')  ['best_breed_prediction'].value_counts()[0:10]

In [ ]:
# Graphic
df_final.query('rating_numerator >= 10')  ['best_breed_prediction'].value_counts()[0:10].plot(kind='bar');
plt.title('Top 10 best breed', size=22)
plt.xlabel('Dog breed')
plt.ylabel('Rating')

plt.plot();

It seems to be the most popular golden retrieriver.